File: ruby/18_4sum.rb
# @param {Integer[]} nums
# @param {Integer} target
# @return {Integer[][]}


# Given array nums = [1, 0, -1, 0, -2, 2], and target = 0.
# [-2, -1, 0, 0, 1, 2]
# A solution set is:
# [
#   [-1,  0, 0, 1],
#   [-2, -1, 1, 2],
#   [-2,  0, 0, 2]
# ]

# 给定数组和一个target，求数组中所有四元组，要求四元组之和为target
# 两次循环，第一层循环从 ai => 0..nums.length-1
# 第二层循环从 di => nums.length-1..ai+3（ai+3，这样两边正好不重叠）
# 然后 bi = ai + 1, ci = di - 1, 又回到了之前的two_sum, 最后将ai，bi，ci，di加入数组存进res即可

def four_sum(nums, target)
    nums = nums.sort

    result = []
    0.upto(nums.length-1-3) do |ai|
    	next if ai > 0 && nums[ai] == nums[ai-1]

    	(nums.length-1).downto(ai+3) do |di|
    		next if di < nums.length-1 && nums[di] == nums[di+1]

    		subsum = target - (nums[ai] + nums[di])
    		bi, ci = ai + 1, di - 1
    		# next unless subsum.between?(nums[bi] * 2, nums[ci] * 2)

    		while bi < ci
    			case nums[bi] + nums[ci] <=> subsum
    			when 1
    				ci -= 1
    			when -1
    				bi += 1
    			when 0
    				item = [nums[ai], nums[bi], nums[ci], nums[di]]
    				result << item #if result.include?(item) == false
    				bi += 1
    				bi += 1 while bi < ci && nums[bi] == nums[bi-1]
    				ci -= 1
    				ci -= 1 while bi < ci && nums[ci] == nums[ci+1]
    			end
    		end
    	end
    end
    result
end

File: cpp/18_4sum.cpp
#include <iostream>
#include <vector>
#include <string>
#include <stack>
#include <queue>
#include <map>
#include <set>
#include <unordered_set>
#include <unordered_map>
using namespace std;

/*
给定一个包含 n 个整数的数组 nums 和一个目标值 target，判断 nums 中是否存在四个元素 a，b，c 和 d ，使得 a + b + c + d 的值与 target 相等？找出所有满足条件且不重复的四元组。
*/

// 首先数组排序，两层循环，然后就是two sum的问题。注意需要过滤重复元素；
class Solution {
public:
    vector<vector<int>> fourSum(vector<int>& nums, int target) {
		vector<vector<int>> res;
		int nsize = nums.size();
		if (nsize == 0) { return res; }
		
		sort(nums.begin(), nums.end());

		for (int i = 0; i < nsize; ++i) {
			int num_1 = nums[i];
			for (int j = i+1; j < nsize; ++j) {
				int num_2 = nums[j];
				// two sum
				int left = j+1; int right = nsize-1;
				while (left < right) {
					int sum = num_1 + num_2 + nums[left] + nums[right];
					if (sum < target) {
						++left;
					} else if (sum > target) {
						--right;
					} else {
						vector<int> item = {num_1, num_2, nums[left], nums[right]};
						res.push_back(item);
						do { ++left; } while (left < right && nums[left] == nums[left-1]);
						do { --right; } while (left < right && nums[right] == nums[right+1]);
					}
				}
				while (j < nsize-1 && nums[j] == nums[j+1]) { ++j; }
			}
			while (i < nsize-1 && nums[i] == nums[i+1]) { ++i; }
		}

		return res;
    }
};
